### PR TITLE
[SPARK-31913][SQL] Fix StackOverflowError in FileScanRDD

### PR DESCRIPTION
**Pull request by uncleGen**
### What changes were proposed in this pull request?

Resolve the recursive calls in `FileScanRDD`.

### Why are the changes needed?
`FileScanRDD` throw a `StackOverflowError` when read a mass of empty files.

There are two ways to scan a file source, i.e. `BucketedRead` and `NonBucketedRead`. When scan a file source in `NonBucketedRead` way, spark will calculate how much files will be read in each partition with two important configuration: `spark.sql.files.maxPartitionBytes` and `spark.sql.files.openCostInBytes`. With these two config, we can calculate the `maxSplitBytes` in each partition. If there are lots of input files and low parallelism, the `maxSplitBytes` will be `spark.sql.files.maxPartitionBytes`. In my failed job, I set `spark.sql.files.maxPartitionBytes` with a large number, like 5GB. It means a spark task will read 5GB files at most. But accidentally, 
a mass of empty files was produced in my environment. Then a spark task will read a mass of empty files. But the `FileScanRDD` read this files in recursive way:
```
     def hasNext: Boolean = {
        // Kill the task in case it has been marked as killed. This logic is from
        // InterruptibleIterator, but we inline it here instead of wrapping the iterator in order
        // to avoid performance overhead.
        context.killTaskIfInterrupted()
        (currentIterator != null && currentIterator.hasNext) || nextIterator()
      }

     private def nextIterator(): Boolean = {
         ...
         try {
           hasNext
         } catch {
         ...
     }
```
`(currentIterator != null && currentIterator.hasNext)` always is `false`, and finally this will trigger a `StackOverflowError`.

!!! **Have to admit**, it is a corner case as  a mass of empty files was produced in my environment accidentally.

```
java.lang.StackOverflowError
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.getSubject(Subject.java:297)
	at org.apache.hadoop.security.UserGroupInformation.getCurrentUser(UserGroupInformation.java:648)
	at org.apache.hadoop.fs.FileSystem$Cache$Key.<init>(FileSystem.java:2828)
	at org.apache.hadoop.fs.FileSystem$Cache$Key.<init>(FileSystem.java:2818)
	at org.apache.hadoop.fs.FileSystem$Cache.get(FileSystem.java:2684)
	at org.apache.hadoop.fs.FileSystem.get(FileSystem.java:373)
	at org.apache.hadoop.fs.Path.getFileSystem(Path.java:295)
	at org.apache.parquet.hadoop.util.HadoopInputFile.fromPath(HadoopInputFile.java:38)
	at org.apache.parquet.hadoop.ParquetFileReader.<init>(ParquetFileReader.java:640)
	at org.apache.spark.sql.execution.datasources.parquet.SpecificParquetRecordReaderBase.initialize(SpecificParquetRecordReaderBase.java:148)
	at org.apache.spark.sql.execution.datasources.parquet.VectorizedParquetRecordReader.initialize(VectorizedParquetRecordReader.java:143)
	at org.apache.spark.sql.execution.datasources.parquet.ParquetFileFormat.$anonfun$buildReaderWithPartitionValues$2(ParquetFileFormat.scala:326)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.org$apache$spark$sql$execution$datasources$FileScanRDD$$anon$$readCurrentFile(FileScanRDD.scala:116)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:169)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:93)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:173)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.hasNext(FileScanRDD.scala:93)
	at org.apache.spark.sql.execution.datasources.FileScanRDD$$anon$1.nextIterator(FileScanRDD.scala:173)
```

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add new UT.

### CODE REVIEW

**Comment by uncleGen** [comment]
pending on fix ut failure

**Comment by uncleGen** [comment]
cc @cloud-fan @xuanyuanking

**Comment by uncleGen** [comment]
retest this please.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileScanRDDSuite.scala [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+import org.apache.spark.sql.QueryTest
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types.{StringType, StructField, StructType}
+
+class FileScanRDDSuite extends QueryTest with SharedSparkSession {
+  import testImplicits._
+
+  test("SPARK-31913: Fix StackOverflowError in FileScanRDD") {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileScanRDDSuite.scala:31]: I run this test in the current master then it passed... Could you check it first?

**Comment by maropu** [comment]
Could you clearly describe what is a root cause of this issue in the PR description?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileScanRDDSuite.scala (hunk 1 above)
Comment (uncleGen) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileScanRDDSuite.scala:31]: I could fail this UT at master branch. Let me check this again.

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileScanRDDSuite.scala:31]: Does this test depend on some env settings?

**Comment by uncleGen** [comment]
@maropu I have updated the PR description, take a review please.

**Comment by xuanyuanking** [comment]
Let me clarify. The issue is the recursive calls in FileScanRDD will cause StackOverflowError while we have too many empty files. Could you please quantify the number of empty files in your env?

**Comment by uncleGen** [comment]
@xuanyuanking Thanks for clarifying. Unfortunately,  those empty files has been removed, I remember there are thousands of empty files at least.

**Comment by uncleGen** [comment]
ping @xuanyuanking

**Comment by xuanyuanking** [comment]
Same question with Takeshi here https://github.com/apache/spark/pull/28737#discussion_r437831211

**Comment by maropu** [comment]
Yea, we need env-independent tests for this issue...

**Pull request closed without merging**